### DISCUSSION TITLE
How to setup a CDK Construct Library for Lambda Function Code Publishing

### DISCUSSION THREAD

**Discussion by nickehunt-idm** [projen/projen#2543] [Q&A]
I have a cdk library project in typescript which works for publishing / consuming all my constructs and classes.  When adding in a lambda class to create a function, I have my function / python code referenced in the construct but is not making it to the lib directory when building so can't be referenced as an asset for the code.  Is there an update to the projenrc / structure I need to add to publish the function code / make it accessible?

Thanks
Nick

**Comment by Chriscbr** [comment]
Hi Nick, we have some docs on how to create lambda functions in AWS CDK project types such as `AwsCdkConstructLibrary` https://projen.io/awscdk.html#aws-lambda-functions - can you check if the information here helps?

**Comment by nickehunt-idm** [comment]
Hey Chris.  Thanks for the quick follow up.  Yeah, saw that and did a quick test to make sure it picks up but is not seeing the .py file when I do a build / package.   am using the .py code for the lambda inline from the lambda.Function construct to create it, using the .py file as the source asset.  I can get everything to test fine referencing the ./src directory location of the file, but the file asset does not make it to the lib / package when doing the full build for my library since is a ts project so my client apps cannot use it when calling my construct from the library...

**Comment by kichik** [comment]
I'm not sure if it's the best way to do it, but you can add a task to copy the files to the `lib` directory. That will get them packaged up. You can see how I do it [here](https://github.com/CloudSnorkel/cdk-github-runners/blob/1e575d17bcde6513ed8ef67f1ee70a5da7b81a14/.projenrc.js#L95-L105).

**Comment by gradybarrett** [comment]
I did this last week and ended up with basically the same solution as kichik. I `rsync`'d the directory with all of my Python lambda runtime artifacts into lib. 

Alternatively, you could add specific `.npmignore` rules such that those files/folders get included during packaging. The downside there is packaged files split between your src and output directories, which then requires you to compensate when using `__dirname` and other relative directives/references.

Might be nice to have an "include" interface of some sort (maybe on NodeProject?) that just does a copy like this during the compile task. `includeUncompiled`, `includeRaw`,  `bundleFiles`, `bundleRaw`? 🤷‍♂️ It seems like a somewhat common use case when using TS because it's not in the compiler's purview to deal with non-TS files.

**Comment by mrgrain** [comment]
This is a documentation issue. All non TS source files should go into a separate directory at the top-level, like this:
```
.
├── lib                      # included, contains the compiled JS files
├── src                     # NOT included, contains the TS sources
├── test                    # NOT included, test files
├── handlers            # included, put your python lambda handlers here
└── ...
```

From your `src/my-lambda.ts` file you can now use a relative path `../handlers/one.py` to reference your python handlers.
This works because while the soruce file ends up in `lib`, it stays in the same relative position to the project root.
